External Metrics support for VPA recommender

lallydd opened the pull request:
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
This adds a mode to the VPA `recommender` to use External Metrics.  It also adds (much more substantially in lines-of-changes) local KIND (Kubernetes IN Docker) support for the `recommender` in both `metrics-server` and External Metrics modes.  

Extending out the testing to support the rest of the VPA tests is certainly possible, and probably 95% done, but this PR was taking a long time already to get sorted.

  
#### Which issue(s) this PR fixes:
Fixes #5153 

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
ALPHA.  Added External Metrics server support.  To enable this mode, and disable the VPA recommender's use 
of `metrics-server` with  `--use-external-metrics=true` and identify the metrics to use for CPU and memory 
recommendations with `--external-metrics-cpu-metric=YOURMETRIC` and 
`--external-metrics-memory-metric=YOURMETRIC`, respectively.  You must use all three flags together to 
configure the VPA in this way.  This is new functionality and is not as mature or as well-tested as the 
existing `metrics-server` functionality.

Will require additional permissions to use:
- apiGroups:
   - "external.metrics.k8s.io"
  resources:
   - "*"
  verbs:
   - get
   - list 
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

voelzmo commented:
Seems like there's some issues with easyCLA: https://github.com/communitybridge/easycla/issues/3843
I contacted sig-contribex about this: https://kubernetes.slack.com/archives/C1TU9EB9S/p1678351853346579

MadhavJivrajani commented:
/reopen 
fingers crossed

MadhavJivrajani commented:
/easycla

voelzmo commented:
/label cncf-cla: no

Trying to manually set the CLA to `no` for now, until upstream issue has been closed.

jarias-lfx commented:
/easycla

voelzmo commented:
yay, thanks for following up here, @nikhita and @jarias-lfx!

sftim commented on vertical-pod-autoscaler/deploy/vpa-rbac.yaml:
```diff
@@ -4,6 +4,18 @@ kind: ClusterRole
 metadata:
   name: system:metrics-reader
```
Is this role name one that SIG Architecture have confirmed? See https://kubernetes.io/docs/reference/access-authn-authz/rbac/#referring-to-subjects for details about reserved names.

lallydd replied:
Where do I check? This line was already here, so I assumed it was OK.

sftim commented on vertical-pod-autoscaler/deploy/vpa-rbac.yaml:
```diff
@@ -4,6 +4,18 @@ kind: ClusterRole
+    resources:
+      - "*"
+    verbs:
+      - "*"
```
This is too much access for a reader role. Only grant read access.

lallydd replied:
Fixing.

sftim commented on vertical-pod-autoscaler/deploy/k8s-metrics-server.yaml:
```diff
@@ -0,0 +1,244 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: system:release-name-metrics-server
```
Is this role name one that SIG Architecture have confirmed? See https://kubernetes.io/docs/reference/access-authn-authz/rbac/#referring-to-subjects for details about reserved names.

lallydd replied:
Ah, this one's a mistake.  I did a simple `helm render` of the original source https://github.com/kubernetes-sigs/metrics-server/blob/master/charts/metrics-server/templates/clusterrole.yaml#L5 but didn't give it a decent name.  I'll update (I'll have a look at the URI you've provided)

sftim commented on vertical-pod-autoscaler/deploy/k8s-metrics-server.yaml:
```diff
@@ -0,0 +1,244 @@
+spec:
+  group: metrics.k8s.io
+  groupPriorityMinimum: 100
+  insecureSkipTLSVerify: true
```
Ouch. I'd expect at least a comment as to why we're doing this.

lallydd replied:
I added a comment to hack/local-cluster.md

lallydd commented:
Squashed as requested.

lallydd commented on vertical-pod-autoscaler/pkg/recommender/Makefile:
```diff
@@ -64,9 +64,9 @@ endif
-	docker manifest create --amend $(REGISTRY)/${FULL_COMPONENT}:$(TAG) $(shell echo $(ALL_ARCHITECTURES) | sed -e "s~[^ ]*~$(REGISTRY)/${FULL_COMPONENT}\-&:$(TAG)~g")
-	@for arch in $(ALL_ARCHITECTURES); do docker manifest annotate --arch $${arch} $(REGISTRY)/${FULL_COMPONENT}:$(TAG) $(REGISTRY)/${FULL_COMPONENT}-$${arch}:${TAG}; done
-	docker manifest push --purge $(REGISTRY)/${FULL_COMPONENT}:$(TAG)
+	docker manifest create --insecure --amend $(REGISTRY)/${FULL_COMPONENT}:$(TAG) $(shell echo $(ALL_ARCHITECTURES) | sed -e "s~[^ ]*~$(REGISTRY)/${FULL_COMPONENT}\-&:$(TAG)~g")
```
Needed for local registry.

jbartosik commented on vertical-pod-autoscaler/pkg/recommender/input/metrics/metrics_source.go:
```diff
@@ -0,0 +1,162 @@
+
+type podContainerResourceMap map[model.PodID]map[string]k8sapiv1.ResourceList
+
+func (s *externalMetricsClient) addMetrics(list *externalmetricsv1beta1.ExternalMetricValueList, name k8sapiv1.ResourceName, resourceMap podContainerResourceMap, namespace, podName string) {
```
I'd change this to

```
func (resourceMap *podContainerResourceMap) addMetrics(list *externalmetricsv1beta1.ExternalMetricValueList, name k8sapiv1.ResourceName, namespace, podName ctrNameLabel, string)
```

since it's adding stuff to podContainerResourceMap. The way it's now I thought it was adding things to `externalMetricsClient` and was very confused why we'd do that

lallydd replied:
Changed.

jbartosik commented on vertical-pod-autoscaler/pkg/recommender/input/metrics/metrics_source.go:
```diff
@@ -0,0 +1,162 @@
+
+		}
+
+		for podId, cmaps := range workloadValues {
```
I don't understand why we copy metric values we got from `nsClient.List()` call to `workloadValues` map then copy from `workloadValues` to `result`. It seems to me that building `result.Items` immediately would be simpler.

And we can just copy `Timestamp` and `Window` for each sample and avoid the problem the comment "I am not sure that this is correct. Need to rework Timestamp and Window... which one should we use?" is talking about

lallydd replied:
You're right.  There was a lot more here before w.r.t. other metircs it was putting out.  I'll merge it all together.

jbartosik commented on vertical-pod-autoscaler/pkg/recommender/input/metrics/metrics_source.go:
```diff
@@ -0,0 +1,162 @@
+type ExternalClientOptions struct {
+	ResourceMetrics                                  map[k8sapiv1.ResourceName]string
+	PodNamespaceLabel, PodNameLabel                  string
+	CtrNamespaceLabel, CtrPodNameLabel, CtrNameLabel string
```
`CtrPodNameLabel` seems unused

jbartosik reviewed (changes requested):
Thanks and sorry for slow review. I think I will be more responsive now.

I left some comments. For now I only looked at changes in vertical-pod-autoscaler/pkg/recommender

lallydd commented on vertical-pod-autoscaler/pkg/recommender/Makefile:
```diff
@@ -64,9 +64,9 @@ endif
-	docker manifest create --amend $(REGISTRY)/${FULL_COMPONENT}:$(TAG) $(shell echo $(ALL_ARCHITECTURES) | sed -e "s~[^ ]*~$(REGISTRY)/${FULL_COMPONENT}\-&:$(TAG)~g")
-	@for arch in $(ALL_ARCHITECTURES); do docker manifest annotate --arch $${arch} $(REGISTRY)/${FULL_COMPONENT}:$(TAG) $(REGISTRY)/${FULL_COMPONENT}-$${arch}:${TAG}; done
-	docker manifest push --purge $(REGISTRY)/${FULL_COMPONENT}:$(TAG)
+	docker manifest create --insecure --amend $(REGISTRY)/${FULL_COMPONENT}:$(TAG) $(shell echo $(ALL_ARCHITECTURES) | sed -e "s~[^ ]*~$(REGISTRY)/${FULL_COMPONENT}\-&:$(TAG)~g")
```
I'll figure out a way to remove it.

lallydd commented on vertical-pod-autoscaler/pkg/recommender/input/metrics/metrics_source.go:
```diff
@@ -0,0 +1,162 @@
+// NewPodMetricsesSource Returns a Source-wrapper around PodMetricsesGetter.
+func NewPodMetricsesSource(source resourceclient.PodMetricsesGetter) PodMetricsLister {
+	return podMetricsSource{metricsGetter: source}
+}
```
It's not, and I'll add documentation explaining that.

lallydd replied:
Yes exactly.

lallydd commented on vertical-pod-autoscaler/pkg/recommender/input/metrics/metrics_source.go:
```diff
@@ -0,0 +1,162 @@
+		var selectedTimestamp v1.Time
+		var selectedWindows time.Duration
+		for resourceName, metricName := range s.options.ResourceMetrics {
+			pods := s.clusterState.GetMatchingPods(vpa)
```
Removing as suggested.

lallydd commented on vertical-pod-autoscaler/pkg/recommender/input/metrics/metrics_source.go:
```diff
@@ -0,0 +1,162 @@
+				if m == nil || len(m.Items) == 0 {
+					continue
+				}
+				klog.V(2).Infof("External Metrics Query for VPA %+v: resource %+v, metric %+v, %d items", vpa.ID, resourceName, metricName, len(m.Items))
```
Changing to V(4).

lallydd commented on vertical-pod-autoscaler/pkg/recommender/input/metrics/metrics_source.go:
```diff
@@ -0,0 +1,162 @@
+type ExternalClientOptions struct {
+	ResourceMetrics                                  map[k8sapiv1.ResourceName]string
+	PodNamespaceLabel, PodNameLabel                  string
+	CtrNamespaceLabel, CtrPodNameLabel, CtrNameLabel string
```
They're almost all removed now -- vestigial remains of instrumentation we've moved out of the `recommender`.   Removed almost all of it.

lallydd commented on vertical-pod-autoscaler/pkg/recommender/input/metrics/metrics_source.go:
```diff
@@ -0,0 +1,162 @@
+type ExternalClientOptions struct {
+	ResourceMetrics                                  map[k8sapiv1.ResourceName]string
+	PodNamespaceLabel, PodNameLabel                  string
+	CtrNamespaceLabel, CtrPodNameLabel, CtrNameLabel string
```
Adding a comment.

lallydd commented on vertical-pod-autoscaler/pkg/recommender/input/cluster_feeder.go:
```diff
@@ -118,7 +118,7 @@ func (m ClusterStateFeederFactory) Make() *clusterStateFeeder {
 
 // NewClusterStateFeeder creates new ClusterStateFeeder with internal data providers, based on kube client config.
```
I'll work with @dbenque to sort this out after whichever one lands first.

jbartosik replied:
Please sync

lallydd commented:
@jbartosik Hi I think I've addressed all your comments.  Do you mind taking another took to see if they work for you?

jbartosik commented on vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1/types.go:
```diff
@@ -40,7 +40,6 @@ type VerticalPodAutoscalerList struct {
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 // +kubebuilder:storageversion
 // +kubebuilder:resource:shortName=vpa
-// +kubebuilder:subresource:status
```
Please don't remove this.

lallydd replied:
It looks like a 'Revert "Add subresource status for vpa"' commit got rolled into a squash.    I'll try a revert-revert and then if that works, squash that.

jbartosik commented on vertical-pod-autoscaler/pkg/recommender/input/metrics/metrics_source.go:
```diff
@@ -0,0 +1,142 @@
+	List(ctx context.Context, namespace string, opts v1.ListOptions) (*v1beta1.PodMetricsList, error)
+}
+
+// podMetricsSource is the metrics-client source of metrics.  When not using External Metrics.
```
"When not using External Metrics." part of the comment seems incomplete

lallydd replied:
Cleaned up.

jbartosik commented on vertical-pod-autoscaler/pkg/recommender/input/metrics/metrics_client.go:
```diff
@@ -47,35 +44,29 @@ type ContainerMetricsSnapshot struct {
 }
 
-func (c *metricsClient) GetContainersMetrics() ([]*ContainerMetricsSnapshot, error) {
+func (c *metricsClient) GetContainersMetrics(state *model.ClusterState) ([]*ContainerMetricsSnapshot, error) {
```
It looks like the `state` argument is not used, let's drop it

lallydd replied:
Removed.

jbartosik replied:
Not yet. Maybe there was some problem with pushing the changes?

jbartosik commented on vertical-pod-autoscaler/pkg/utils/vpa/api.go:
```diff
@@ -49,14 +49,14 @@ type patchRecord struct {
 }
 
-func patchVpaStatus(vpaClient vpa_api.VerticalPodAutoscalerInterface, vpaName string, patches []patchRecord) (result *vpa_types.VerticalPodAutoscaler, err error) {
+func patchVpa(vpaClient vpa_api.VerticalPodAutoscalerInterface, vpaName string, patches []patchRecord) (result *vpa_types.VerticalPodAutoscaler, err error) {
```
Looks like this is reverting some changes, please avoid this

lallydd commented:
Pushed now, sorry for delay.

lallydd commented:
(I see the failure, working on it now)

kwiesmueller commented on vertical-pod-autoscaler/pkg/recommender/input/metrics/metrics_source.go:
```diff
@@ -0,0 +1,143 @@
+					return nil, err
+				}
+				if m == nil || len(m.Items) == 0 {
+					continue
```
Would it make sense to add a debug log here too indicating that we got an empty response?

kwiesmueller commented on vertical-pod-autoscaler/pkg/recommender/main.go:
```diff
@@ -122,14 +127,30 @@ func main() {
+
 	// CappingPostProcessor, should always come in the last position for post-processing
 	postProcessors = append(postProcessors, &routines.CappingPostProcessor{})
+	var source input_metrics.PodMetricsLister = nil
```
Is the `= nil` needed?
I think it should be nil without that too?

jbartosik commented on vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1/types.go:
```diff
@@ -40,6 +40,7 @@ type VerticalPodAutoscalerList struct {
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 // +kubebuilder:storageversion
 // +kubebuilder:resource:shortName=vpa
+// +kubebuilder:subresource:status
```
This looks like it's part of #5766

jbartosik commented on vertical-pod-autoscaler/pkg/recommender/input/metrics/metrics_client_test_util.go:
```diff
@@ -85,7 +85,7 @@ func (tc *metricsClientTestCase) createFakeMetricsClient() MetricsClient {
 		return true, tc.getFakePodMetricsList(), nil
 	})
-	return newMetricsClient(fakeMetricsGetter.MetricsV1beta1(), "", "fake")
+	return NewMetricsClient(NewPodMetricsesSource(fakeMetricsGetter.MetricsV1beta1()), "", "client")
```
I'd rather keep "fake" as client name

jbartosik commented on vertical-pod-autoscaler/pkg/utils/vpa/api.go:
```diff
@@ -49,14 +49,14 @@ type patchRecord struct {
 }
 
-func patchVpa(vpaClient vpa_api.VerticalPodAutoscalerInterface, vpaName string, patches []patchRecord) (result *vpa_types.VerticalPodAutoscaler, err error) {
+func patchVpaStatus(vpaClient vpa_api.VerticalPodAutoscalerInterface, vpaName string, patches []patchRecord) (result *vpa_types.VerticalPodAutoscaler, err error) {
```
Changes in this file look like they're part of #5766 too

lallydd replied:
I kept two commits here: 1 is a revert for a commit that somehow fell into this branch, and another for all of my changes.  I'll try rebasing and see if I can skip the revert commit I added.

jbartosik reviewed (commented):
Source code changes look mostly good to me, I left a few small comments, @kwiesmueller also left some

jbartosik commented on vertical-pod-autoscaler/deploy/k8s-metrics-server.yaml:
Why do we need this? I think it's better to refer people to https://github.com/kubernetes-sigs/metrics-server/tree/master than to try and host our own copy

jbartosik commented on vertical-pod-autoscaler/deploy/metrics-pump.yaml:
Do we need this? I'd prefer to refer people to a place where manifest will be actively maintained (same for `prometheus-adapter.yaml`, `prometheus.yaml`, )

jbartosik commented on vertical-pod-autoscaler/deploy/recommender-externalmetrics-deployment.yaml:
```diff
@@ -0,0 +1,42 @@
+          - --use-external-metrics=true
+          - --external-metrics-cpu-metric=cpu
+          - --external-metrics-memory-metric=mem
+          - --v=4
```
Why increased verbosity?

jbartosik commented on vertical-pod-autoscaler/deploy/vpa-rbac.yaml:
```diff
@@ -140,6 +165,19 @@ subjects:
 ---
 apiVersion: rbac.authorization.k8s.io/v1
 kind: ClusterRoleBinding
+metadata:
```
Looks like some more changes from #5766

jbartosik commented on vertical-pod-autoscaler/deploy/vpa-rbac.yaml:
```diff
@@ -4,6 +4,20 @@ kind: ClusterRole
 metadata:
   name: system:metrics-reader
 rules:
+  - apiGroups:
```
We need a release note about changing permissions.

Also I expect many VPA instances will not need those permissions. What do you think about not giving those permissions unless necessary

lallydd replied:
I thought that the `vertical-pod-autoscaler/deploy/*` config were all just to support testing.  The actual deployment configurations should be under the top-level `charts/` subdirectory.  Is that wrong?

lallydd replied:
I think leaving the permissions out unless necessary is the right thing to do.  You can't configure a recommender with external metrics without doing some of your own configuration (e.g., flags for the names of the metrics to use for cpu and memory).  I think when mentioning external metrics support in the release notes, we should also mention the flags and additional permissions required.

lallydd replied:
I've reduced them to just `externalmetrics` now, and will move them over now to another file.

jbartosik commented on vertical-pod-autoscaler/deploy/vpa-v1-crd-gen.yaml:
```diff
@@ -513,7 +513,8 @@ spec:
     served: true
     storage: true
-    subresources: {}
+    subresources:
```
Looks like this is from #5766

jbartosik commented on vertical-pod-autoscaler/e2e/go.sum:
Changes to `vertical-pod-autoscaler/e2e` seem unnecessary

lallydd replied:
Moved to `hack/e23`

jbartosik reviewed (changes requested):
I took a look at changes outside `recommender` dir too

lallydd commented on vertical-pod-autoscaler/hack/deploy-for-e2e-locally.sh:
```diff
@@ -0,0 +1,82 @@
+#!/bin/bash
+
+# Copyright 2018 The Kubernetes Authors.
```
Changed (locally, you'll see them on my next push after I rebase & test).

lallydd commented on vertical-pod-autoscaler/hack/emit-metrics.py:
```diff
@@ -0,0 +1,150 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import argparse
```
For this and the other comments, I'm going to reply with a thumbs-up to mean "Changed (locally, you'll see them on my next push after I rebase & test)."

jbartosik replied:
Can we use https://pkg.go.dev/k8s.io/kubernetes/test/images/resource-consumer#section-readme instead?

jbartosik replied:
Please open a follow up issue to make this more maintainable

lallydd commented:
@kwiesmueller @jbartosik anything else?  I believe the rebase cleared up the #5766 issues.  I made the changes mentioned in the `recommender` dir, and the `deploy/` and `e2e/` changes are to support local `kind`-based testing.  They shouldn't affect the production configs that (AFAIK) are under the top-level `charts/` subdir.

lallydd commented on vertical-pod-autoscaler/hack/emit-metrics.py:
```diff
@@ -0,0 +1,156 @@
+# This script runs as a pod, scanning the cluster for other pods.
+# It then publishes fake metrics (Gaussian by --mean_* and --stddev_*)
+# for each pod into a Prometheus Pushgateway.  The Prometheus instance
+# connected to that Pushgateway can have a Prometheus Adapter connected
```
The Push gateway is slightly less complex than setting this up as an HTTP endpoint. Not a lot, but a little.

jbartosik commented on vertical-pod-autoscaler/pkg/recommender/main.go:
```diff
@@ -122,14 +127,30 @@ func main() {
+		klog.V(1).Infof("Using External Metrics: %+v", externalClientOptions)
+		source = input_metrics.NewExternalClient(config, clusterState, *externalClientOptions)
+	} else {
+		source = input_metrics.NewPodMetricsesSource(resourceclient.NewForConfigOrDie(config))
```
Please log here, similar to what we log in line 143; so we don't have to guess that we're using metrics from metrics server by absence of log about using external metrics.

lallydd replied:
Done.

jbartosik reviewed (commented):
VPA for e2e is deployed by [deploy-for-e2e.sh](https://github.com/kubernetes/autoscaler/blob/265e57c1636298729002695326db32205d866f48/vertical-pod-autoscaler/hack/deploy-for-e2e.sh#L1) script from yaml files in https://github.com/kubernetes/autoscaler/tree/master/vertical-pod-autoscaler/deploy.

I think we could modify the script to run recommender e2e on recommender deployed with a different metrics source.

jbartosik commented:
> [Kevin Wiesmüller](https://github.com/kwiesmueller) [Joachim](https://github.com/jbartosik) are the code changes under recommender/ otherwise OK?

I don't see any problems other than the extra log I asked for.

lallydd commented on cluster-autoscaler/cloudprovider/aws/aws_sdk_provider.go:
```diff
@@ -61,11 +63,14 @@ func createAWSSDKProvider(configReader io.Reader) (*awsSDKProvider, error) {
+	// add cluster-autoscaler to the user-agent to make it easier to identify
+	agent := fmt.Sprintf("cluster-autoscaler/v%s", version.ClusterAutoscalerVersion)
+	sess.Handlers.Build.PushBack(request.MakeAddToUserAgentFreeFormHandler(agent))
+
```
I think a rebase should fix this.

lallydd replied:
Yup, AFAICT, the file isn't even in the diff anymore.

lallydd commented:
I rebased to see if that fixes the failed tests in cluster autoscaler.

x13n commented:
/remove-label cluster-autoscaler

x13n commented:
/shrug

lallydd commented:
Tests pass after another rebase.

voelzmo reviewed (changes requested):
In the current state, the local e2e tests fail with this in the recommender logs 
```
Cannot update VPA vertical-pod-autoscaling-3092/hamster-vpa object. Reason: verticalpodautoscalers.autoscaling.k8s.io "hamster-vpa" is forbidden: User "system:serviceaccount:kube-system:vpa-recommender" cannot patch resource "verticalpodautoscalers/status" in API group "autoscaling.k8s.io" in the namespace "vertical-pod-autoscaling-3092"
```

Now that https://github.com/kubernetes/autoscaler/pull/5911 has been merged, the permissions in `hack/e2e/vpa-rbac.yaml` need to be adapted as well. 

This is a bit brittle, people will have to remember to adapt permissions in two places, because it is duplicated. Not sure if there's a good alternative, though.

On a more general note: I really like adding the ability to run the e2e tests locally in a kind cluster! If we get all tests to run in this scenario, this would be very helpful to everyone developing the VPA! I just really wish you would have PR'ed this separately. Right now this is a big chunk with different concerns and hard to review/approve (at least for me, it is).

lallydd commented:
@voelzmo It worked!  Now this is just the diff necessary: https://github.com/kubernetes/autoscaler/pull/5576/commits/68c9db88f870c336499e027c9f73882be2fdf8e2#diff-debb7e710c67f3afe14c9605ad0d7ab3b59f6272a15e21d55679c27b139d732d

And that's all just to enable external metrics, so it's a useful artifact to maintain generally.

lallydd commented:
Also added permissions note to the `release-notes` PR header.  Please LMK if that's not the right place for it.

jbartosik commented on vertical-pod-autoscaler/hack/deploy-for-e2e-locally.sh:
```diff
@@ -0,0 +1,86 @@
+kubectl apply -f ${SCRIPT_ROOT}/hack/e2e/k8s-metrics-server.yaml
+
+for i in ${COMPONENTS}; do
+  if [ $i == admission-controller ] ; then
```
If I understand correctly this should never happen. Check in line 43 ensures that `COMPONENTS` is `recommender` or `recommender-externalmetrics`.

jbartosik commented on vertical-pod-autoscaler/hack/deploy-for-e2e-locally.sh:
```diff
@@ -0,0 +1,86 @@
+  echo "<suite> should be one of:"
+  echo " - recommender"
+  echo " - recommender-externalmetrics"
+  echo "If component is not specified all above will be started."
```
This doesn't seem to be true. If there is no component specified script prints help and exits

jbartosik commented on vertical-pod-autoscaler/hack/deploy-for-e2e-locally.sh:
```diff
@@ -0,0 +1,86 @@
+esac
+
+# Local KIND-hosted registry
+export REGISTRY=localhost:5001
+export TAG=latest
```
I think it would be nice if the script allowed setting REGISTRY and TAG

jbartosik commented on vertical-pod-autoscaler/hack/deploy-for-e2e-locally.sh:
```diff
@@ -0,0 +1,86 @@
+kubectl apply -f ${SCRIPT_ROOT}/deploy/vpa-v1-crd-gen.yaml
+kubectl apply -f ${SCRIPT_ROOT}/hack/e2e/k8s-metrics-server.yaml
+
+for i in ${COMPONENTS}; do
+  if [ $i == admission-controller ] ; then
+    (cd ${SCRIPT_ROOT}/pkg/${i} && bash ./gencerts.sh || true)
+  elif [ $i == recommender-externalmetrics ] ; then
+    i=recommender
+  fi
+  ALL_ARCHITECTURES=amd64 make --directory ${SCRIPT_ROOT}/pkg/${i} release REGISTRY=${REGISTRY} TAG=${TAG}
+  kind load docker-image ${REGISTRY}/vpa-${i}-amd64:${TAG}
+done
```
```suggestion
  kind load docker-image ${REGISTRY}/vpa-recommender-amd64:${TAG}
```

jbartosik replied:
No, it should stay

jbartosik commented on vertical-pod-autoscaler/hack/e2e/k8s-metrics-server.yaml:
I think I've asked about this but can't find the comment right now: is this a copy of a file from another repo?

jbartosik commented on vertical-pod-autoscaler/hack/e2e/vpa-rbac.diff:
Please explain what this diff is for

jbartosik commented on vertical-pod-autoscaler/hack/run-e2e-locally.sh:
```diff
@@ -0,0 +1,70 @@
+    ${SCRIPT_ROOT}/hack/deploy-for-e2e-locally.sh ${SUITE}
+
+    echo " ** Running suite ${SUITE}"
+    if [ ${SUITE} == recommender-externalmetrics ]; then
+       ${SCRIPT_ROOT}/hack/run-e2e-tests.sh recommender
+    else
+      ${SCRIPT_ROOT}/hack/run-e2e-tests.sh ${SUITE}
+    fi
```
```suggestion
    ${SCRIPT_ROOT}/hack/run-e2e-tests.sh recommender
```

jbartosik commented on vertical-pod-autoscaler/hack/vpa-process-yamls.sh:
```diff
@@ -40,7 +40,19 @@ if [ $# -gt 2 ]; then
+function script_path {
+  # Regular components have deployment yaml files under /deploy/.  But some components only have
+  # test deployment yaml files that are under hack/e2e.
+  if printf '%s\0' "${TEST_COMPONENTS[@]}" | grep -Fqxz -- $1; then
```
I don't understand this condition. Could you explain it (or rewrite it to make it easier to understans)

jbartosik reviewed (changes requested):
I took a look at parts of the PR other than the core code change. Mostly they look good but I'm asking for some changes (should be small) or examplations.

Thank you!

lallydd reviewed (commented):
I believe I got everything, except for replacing the script as we discusssed in this morning's zoom call.

jbartosik commented on vertical-pod-autoscaler/e2e/go.sum:
Please remove changes in vertical-pod-autoscaler/e2e/go.sum and vertical-pod-autoscaler/e2e/vendor/ from the PR

lallydd replied:
Done and done.

jbartosik commented on vertical-pod-autoscaler/vendor/k8s.io/metrics/pkg/apis/external_metrics/doc.go:
There aren't any changes to go.mod / go.sum causing this?

lallydd replied:
The `external_metrics` api was referenced in vendor/modules.txt.  I think it's just pulled in as part of the existing `k8s.io/metrics 0.25.0` reference in `go.mod`

lallydd replied:
Ref: https://github.com/kubernetes/autoscaler/commit/a5148d2a10892f02ab724d533bd6151590d69d46#diff-34952f46f28ac4a379ca6d8979d2d2acff269f9b50db916304f99a5ae556e5d1R653-R656

jbartosik reviewed (commented):
Mostly looks good to me. A few comments to clean this up.

jbartosik reviewed (commented):
/lgtm

As discussed at the last SIG meeting we should clearly document flags we add as alpha and communicate clearly that this isn't as mature as metrics server support yet.

https://kubernetes.io/docs/reference/using-api/deprecation-policy/#deprecating-a-flag-or-cli

lallydd commented:
@jbartosik I added ALPHA designations to the command-line flags and added an ALPHA designation to the release notes.  Also this sentence: "This is new functionality and is not as mature or as well-tested as the 
existing `metrics-server` functionality."

jbartosik reviewed (approved):
/lgtm
/approve
